backend/remote-state/oss/client.go

```diff
@@ -0,0 +1,232 @@
+	acl                  string
+}
+
+func (c *RemoteClient) Get() (payload *remote.Payload, err error) {
+	buf, err := c.getObj(c.statePath)
+	if err != nil {
+		return nil, err
+	}
+
+	// If there was no data, then return nil
+	if buf == nil || len(buf.Bytes()) == 0 {
+		log.Printf("State %s has no data.", c.statePath)
+		return nil, nil
+	}
+
+	var hashChannel = make(chan []byte, 1)
+	sum := md5.Sum(buf.Bytes())
+	hashChannel <- sum[:]
+	payload = &remote.Payload{
+		Data: buf.Bytes(),
+		MD5:  <-hashChannel,
+	}
+	return payload, nil
+}
+
+func (c *RemoteClient) Put(data []byte) error {
+	return c.putObj(c.statePath, data)
+}
+
+func (c *RemoteClient) Delete() error {
+	return c.deleteObj(c.statePath)
+}
+
+func (c *RemoteClient) Lock(info *state.LockInfo) (string, error) {
+	bucket, err := c.ossClient.Bucket(c.bucketName)
+	if err != nil {
+		return "", fmt.Errorf("Error getting bucket: %#v", err)
+	}
+
+	log.Printf("Lock info:%#v", info)
+
+	infoJson, err := json.Marshal(info)
+	if err != nil {
+		return "", err
+	}
+
+	if info.ID == "" {
+		lockID, err := uuid.GenerateUUID()
+		if err != nil {
+			return "", err
+		}
+		info.ID = lockID
+	}
+	info.Path = c.lockPath
+	if exist, err := bucket.IsObjectExist(info.Path); err != nil {
+		return "", fmt.Errorf("Estimating object %s is exist got an error: %#v", info.Path, err)
+	} else if !exist {
+		if err := c.putObj(info.Path, infoJson); err != nil {
+			return "", err
+		}
+	} else if _, err := c.validLock(info.ID); err != nil {
+		return "", err
+	}
+
+	return info.ID, nil
+}
+
+func (c *RemoteClient) Unlock(id string) error {
+	log.Printf("UnLock info %s", id)
+	lockInfo, err := c.validLock(id)
+	if err != nil {
+		return err
+	}
+
+	if err := c.deleteObj(c.lockPath); err != nil {
+		return &state.LockError{
+			Info: lockInfo,
+			Err:  err,
+		}
+	}
+	return nil
+}
+
+func (c *RemoteClient) putObj(key string, data []byte) error {
+	log.Printf("Put Object %s.", key)
+	bucket, err := c.ossClient.Bucket(c.bucketName)
+	if err != nil {
+		return fmt.Errorf("Error getting bucket: %#v", err)
+	}
+	body := bytes.NewReader(data)
+
+	var options []oss.Option
+	if c.acl != "" {
+		options = append(options, oss.ACL(oss.ACLType(c.acl)))
+	}
+	options = append(options, oss.ContentType("application/json"))
+	if c.serverSideEncryption {
+		options = append(options, oss.ServerSideEncryption("AES256"))
+	}
+	options = append(options, oss.ContentLength(int64(len(data))))
+
+	if body != nil {
+		if err := bucket.PutObject(key, body, options...); err != nil {
+			return fmt.Errorf("failed to upload %s: %#v", key, err)
+		}
+		log.Printf("Put Object %s successfully.", key)
+		return nil
+	}
+	return nil
+}
+
+func (c *RemoteClient) getObj(key string) (*bytes.Buffer, error) {
+	log.Printf("Get Object %s.", key)
+	bucket, err := c.ossClient.Bucket(c.bucketName)
+	if err != nil {
+		return nil, fmt.Errorf("Error getting bucket: %#v", err)
+	}
+
+	if exist, err := bucket.IsObjectExist(key); err != nil {
```

Comment:
Do we really need to check existence first?
This of course introduces the possibility of the object being deleted before the GetObject call.